File: client/services/authService.ts
import api from "@/lib/api";
import { ApiResponse } from "@/services/constants/type";

// ---------- Types ----------

export type RegisterRequest = {
  username: string;
  password: string;
  email: string;
};

export type LoginRequest = {
  email: string;
  password: string;
};

export type LoginResponse = {
  accessToken: string;
  expiresAt: number;
  user: {
    id: string;
    email: string;
    username: string;
    fullname: string;
    avatarUrl: string | null;
    roles: string[];
  };
};
export interface RegisterResponse {
  id: string;
  email: string;
  urlToken: string;
  createdAt: string;
}

export type RefreshResponse = LoginResponse;

export const AuthService = {
  async register(
    payload: RegisterRequest,
  ): Promise<ApiResponse<RegisterResponse>> {
    const { data } = await api.post<ApiResponse<RegisterResponse>>(
      "/auth/register",
      payload,
    );
    return data;
  },
  async activate(urlToken: string, activateCode: string) {
    await api.post<ApiResponse<null>>("/auth/activate", {
      urlToken,
      activateCode,
    });
  },

  async login(payload: LoginRequest) {
    const  data  = await api.post<ApiResponse<LoginResponse>>(
      "/auth/login",
      payload,
    );
    return data;
  },

  async refresh(): Promise<ApiResponse<RefreshResponse>> {
    const { data } =
      await api.post<ApiResponse<RefreshResponse>>("/auth/refresh");
    return data;
  },
  async loginGoogle(idToken: String): Promise<ApiResponse<LoginResponse>> {
    const { data } = await api.post<ApiResponse<LoginResponse>>(
      "auth/social/google",
      { idToken },
    );
    return data;
  },

  async logout(): Promise<void> {
    await api.post("/auth/logout");
  },
};

File: server/src/main/java/com/web/TradeApp/feature/user/auth/service/AuthService.java
package com.web.TradeApp.feature.user.auth.service;

import com.web.TradeApp.feature.user.auth.dto.LoginRequest;
import com.web.TradeApp.feature.user.auth.dto.LoginResponse;
import com.web.TradeApp.feature.user.auth.dto.RegisterRequest;
import com.web.TradeApp.feature.user.auth.dto.RegisterResponse;
import com.web.TradeApp.feature.user.entity.User;

import jakarta.mail.MessagingException;

public interface AuthService {
    RegisterResponse register(RegisterRequest request);

    void activateAccount(String codeRequest, String urlTokenRequest) throws MessagingException;

    LoginResponse login(LoginRequest request);

    LoginResponse createLoginRes(User user, String email);
}
